### ISSUE TITLE
How to achieve install of Kilo on k3s cluster, behind NAT?

### ISSUE THREAD

**Issue by tetricky** [squat/kilo#76]
Opening this as a new issue, as per request. See also https://github.com/squat/kilo/issues/27

k3s (v1.18.8+k3s1) on debian buster with iptables (and ufw). Single master, two worker nodes (KVM vm's networked with NAT, can reach each other on the subnet, 6443 and 51820 open on all nodes). No traefik, metallb, or flannel (I'm aiming to get a bare cluster, based on kilo networking). The kilo manifest points to "kubeconfig=/etc/kubernetes/kubeconfig" (from the kilo-k3s.yaml). Which doesn't exist on any node. On the server kubeconfig is at /etc/rancher/k3s/k3s.yaml the server pod runs, but log has:

{"caller":"cni.go:58","component":"kilo","err":"failed to read IPAM config from CNI config list file: no IP ranges specified","level":"warn","msg":"failed to get CIDR from CNI file; overwriting it","ts":"2020-09-14T18:52:31.564522566Z"}
{"caller":"cni.go:66","component":"kilo","level":"info","msg":"CIDR in CNI file is empty","ts":"2020-09-14T18:52:31.56461313Z"}
{"CIDR":"10.42.0.0/24","caller":"cni.go:71","component":"kilo","level":"info","msg":"setting CIDR in CNI file","ts":"2020-09-14T18:52:31.564666128Z"}

On the worker nodes /etc/rancher/k3s/k3s.yaml appears to be an empty directory. and the pods running on the nodes are "CrashLoopBackOff" and log has:

failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": read /etc/kubernetes/kubeconfig: is a directory

I'm working off https://jbhannah.net/articles/k3s-wireguard-kilo with some mods (addition of iptables to debian buster, etc).

The suggestion (thank you very much) was:
"regarding the problem you're experiencing, please take note of the comment in the k3s manifest: https://github.com/squat/kilo/blob/master/manifests/kilo-k3s.yaml#L169-L172

We need to copy the k3s.yaml file from the master to the same place on the worker nodes. In your case, that means you should:

    delete the Kilo daemonset;
    delete the /etc/rancher/k3s/k3s.yaml from the worker nodes;
    make sure the k3s.yaml file from the master node uses a server address that is routable from the internet, e.g. an externally resolvable DNS name that resolves to a public IP address;
    copy the file to the worker nodes; and then
    reapply the kilo manifests to the cluster."

**Comment by tetricky** [comment]
This I have followed (to the best of my ability). Not with an externally routeable ip, but within the NAT subnet.  I have copied /etc/rancher/k3s/k3s.yaml from the server to the workers, and tried applying the manifest with both 127.0.0.1 and the server ip address (192.168.122.101) as the kubeconfig target.

These are the logs (the workers are still CrashLoopBackOff). Is this indicative of a deeper problem (Certificate Authority?), or am I still failing to understand how to get this to work (I mean, clearly yes...but how stupid am I being)?

root@k3s-server:~# kubectl logs -f kilo-xkn8m -n kube-system

{"caller":"main.go:217","msg":"Starting Kilo network mesh 'ab8df1306eff509b716d5282e1fbc5df8760011c'.","ts":"2020-09-15T10:36:58.552852456Z"}
{"caller":"cni.go:58","component":"kilo","err":"failed to read IPAM config from CNI config list file: no IP ranges specified","level":"warn","msg":"failed to get CIDR from CNI file; overwriting it","ts":"2020-09-15T10:36:58.654093469Z"}
{"caller":"cni.go:66","component":"kilo","level":"info","msg":"CIDR in CNI file is empty","ts":"2020-09-15T10:36:58.654164388Z"}
{"CIDR":"10.42.0.0/24","caller":"cni.go:71","component":"kilo","level":"info","msg":"setting CIDR in CNI file","ts":"2020-09-15T10:36:58.654197571Z"}

root@k3s-server:~# kubectl logs -f kilo-pz4ck -n kube-system

failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": v1.Config.Contexts: []v1.NamedContext: Clusters: []v1.NamedCluster: v1.NamedCluster.Name: Cluster: v1.Cluster.Server: CertificateAuthorityData: decode base64: illegal base64 data at input byte 161, error found in #10 byte of ...|EFlRncwe$","server":|..., bigger context ...|3kxelpYSjIKWlhJdFkyRkFNVFl3TURFd01qQTBPREFlRncwe$","server":"https://192.168.122.101:6443"},"name":"|...

root@k3s-server:~# kubectl logs -f kilo-spxsc -n kube-system

failed to create Kubernetes config: Error loading config file "/etc/kubernetes/kubeconfig": v1.Config.Contexts: []v1.NamedContext: Clusters: []v1.NamedCluster: v1.NamedCluster.Name: Cluster: v1.Cluster.Server: CertificateAuthorityData: decode base64: illegal base64 data at input byte 161, error found in #10 byte of ...|EFlRncwe$","server":|..., bigger context ...|3kxelpYSjIKWlhJdFkyRkFNVFl3TURFd01qQTBPREFlRncwe$","server":"https://192.168.122.101:6443"},"name":"|...

**Comment by squat** [comment]
Hi, thanks for starting this issue :) this looks like something is corrupt in the kubeconfig you copied from one node to the other. Specifically, the Kilo Pod on the master is starting perfectly fine but only the Pods on the worker nodes are failing, so there _must_ be a difference in the file. Can you please share a diff of the files on each node? I think that would help us track down the problem.

**Comment by tetricky** [comment]
I think you are on to something. I cut and pasted the text of the file, and I think (now you have said. D'oh.  Schoolboy error) that might have truncated the secret. I'll re-run it actually copying kubeconfig, not cut and pasting. Will let you know how it goes tomorrow.

Great catch, told you I was an idiot.  To follow...

**Comment by tetricky** [comment]
To complete this issue.

I copied the kubeconfig from the master node (using rysnc), to the two worker nodes. (In k3s this being /etc/rancher/k3s/k3s.yaml). I changed the "server:" ip to that of the (NAT subnet) master node.

I then reapplied the kilo manifests to the cluster, using kubectl from the master node.

Everything appears to have come up now. "wg" on each of the nodes shows the interface and the other peers correctly connected, and the kilo pod logs are showing handshaking with the other (respective) peers.

So, as suggested my error in non-rigourously copying the kubeconfig correctly. A simple basic error.

So I'm marking this issue as closed, and just need to learn how to use it now!
